In [78]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [79]:
dataFrame = pd.read_csv('creditcard.csv')
dataFrame


Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

In [80]:
#1 -> fraudulent
#0 -> not fraudulent
dataFrame.Class.value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [81]:
#Porcentagem de fraudes
fraudes = dataFrame.Class.value_counts()[1]
nao_fraude = dataFrame.Class.value_counts()[0]
total = len(dataFrame)

print('Porcentagem de fraudes: {0:.2f}%'.format((fraudes/total)*100))
print('Porcentagem de não fraudes: {0:.2f}%'.format((nao_fraude/total)*100))

Porcentagem de fraudes: 0.17%
Porcentagem de não fraudes: 99.83%


<h4>Separando X e Y</h4>

In [82]:
from sklearn.model_selection import train_test_split

SEED = 1231
np.random.seed(SEED)

raw_train_x, raw_test_x, train_y, test_y = train_test_split(dataFrame.drop('Class', axis = 'columns'), dataFrame['Class'], test_size=0.25)

<h2>Função para imprimir os resultados</h2>

In [83]:
def imprimeResult(results):
    media = results['test_score'].mean()
    print("Resultados cross validation:",results['test_score'])
    print("Média da precisão: {0:.2f}".format(media*100),'%')

    #Acurracy fica entre a média + 2(desvio_padrao) e a média - 2(desvio_padrao)
    desvio_padrao = results['test_score'].std()
    print("Desvio padrão: {0:.2f}".format(desvio_padrao*100),'%')

    acumulo_baixo = media - 2*desvio_padrao
    acumulo_alto = media + 2*desvio_padrao
    print("Acurracy esta entre: [",acumulo_baixo*100,"---", acumulo_alto*100,"]")

<h1>Linear SVC</h1>

In [84]:
# from sklearn.svm import LinearSVC
# from sklearn.model_selection import cross_validate
# from sklearn.model_selection import StratifiedKFold

# model = LinearSVC()

# cv = StratifiedKFold(n_splits=5, shuffle=True)

# results = cross_validate(model, raw_train_x, train_y, cv = cv, return_train_score = False)

# imprimeResult(results)

<h3>A precisão é alta mas, é possível notar que o modelo está treinando com poucas informações de fraude. (UnderSampling)</h3><br>

<h1> É necessário realizar um treino com um campo de teste mais 'justo'</h1>
<p> Para isso, é necessário reduzir os casos 'não fraudados' para 492 também (50/50) </p> 
<p> Um problema é que o modelo irá treinar com menos dados </p>

In [85]:
#Pegando uma amostra de dados para testar o modelo

amostra_dataFrame = dataFrame.sample(frac = 1)  #frac = 1 pega todos os dados

fraude = amostra_dataFrame[amostra_dataFrame.Class == 1]
n_fraude = amostra_dataFrame[amostra_dataFrame.Class == 0][:492] #amostra de 492 casos não fraudados para o treino do modelo


In [86]:
#Juntando as amostras para o treino
dataFrame_justo = pd.concat([fraude, n_fraude])
dataFrame_justo

Time         V1        V2         V3        V4         V5  \
46998    43028.0  -1.109646  0.811069  -1.138135  0.935265  -2.330248   
42936    41305.0 -12.980943  6.720508 -13.455636  8.698610 -11.479552   
274382  165981.0  -5.766879 -8.402154   0.056543  6.950983   9.880564   
30442    35926.0  -3.896583  4.518355  -4.454027  5.547453  -4.121459   
114271   73408.0  -2.869795  1.335667  -1.009530  1.693885  -0.741480   
...          ...        ...       ...        ...       ...        ...   
152216   97109.0  -0.533261  1.650537   0.040631  0.996911   1.958309   
115926   74074.0  -1.328804  0.733208   0.987400  1.613179  -1.141933   
224872  144004.0  -0.811926  0.162793   0.450945  0.286060   1.081884   
223852  143550.0   2.032448 -0.022721  -1.429822  0.132413   0.351459   
59221    48747.0   1.202207  0.345933   0.288115  0.640528  -0.118152   

              V6         V7        V8        V9  ...       V21       V22  \
46998  -0.116106  -1.621986  0.458028 -0.912189  ...  0.641594  0.841755   
42936  -2.681519 -14.019291  8.218191 -7.930900  ...  2.549628 -0.532228   
274382 -5.773192  -5.748879  0.721743 -1.076274  ...  0.880395 -0.130436   
30442  -1.163407  -6.805053  2.928356 -4.917130  ...  1.691042  0.920021   
114271 -0.796773  -2.614242  1.066636 -1.135497  ...  0.528421  0.228027   
...          ...        ...       ...       ...  ...       ...       ...   
152216  0.122127   1.327374 -0.129318 -0.144995  ... -0.148206 -0.083901   
115926  1.443504  -0.364173  1.241906  0.547513  ... -0.288443 -0.673991   
224872 -0.934992   0.793938 -0.332843 -0.281152  ... -0.123789 -0.098608   
223852 -0.541265   0.121327 -0.170956  0.236004  ...  0.304590  1.006161   
59221  -0.581460   0.072089 -0.112992 -0.095121  ... -0.247013 -0.652064   

             V23       V24       V25       V26       V27       V28  Amount  \
46998   0.176728  0.081004 -0.258899  0.707654  0.418649  0.080756  204.27   
42936  -0.235096  0.673209  0.226598 -0.006168 -1.185696 -0.747361   59.68   
274382  2.241471  0.665346 -1.890041 -0.120803  0.073269  0.583799    0.00   
30442  -0.151104  0.011007  0.080303  0.412191  0.635789  0.501050    4.56   
114271 -0.749298 -0.067179  0.215792 -0.453651  0.349191  0.198862   16.48   
...          ...       ...       ...       ...       ...       ...     ...   
152216 -0.648335  0.064670  1.153640 -0.248047  0.026864  0.070300   10.03   
115926  0.160231 -0.807054 -0.265240 -0.383635 -0.080586 -0.074501  118.76   
224872 -0.041533  1.204010 -0.555049 -0.509975  0.220713  0.145950   85.76   
223852  0.027098  0.756044  0.257735 -0.160604 -0.021378 -0.060979    0.89   
59221   0.143825  0.056505  0.168621  0.120756 -0.002399  0.030198    7.76   

        Class  
46998       1  
42936       1  
274382      1  
30442       1  
114271      1  
...       ...  
152216      0  
115926      0  
224872      0  
223852      0  
59221       0  

[984 rows x 31 columns]

984 linhas, sendo metade delas fraude e a outra metade não fraudadas.

<h1>Linear SVC com dados justos</h1>

In [87]:
from sklearn.svm import LinearSVC
from sklearn.model_selection import cross_validate
from sklearn.model_selection import StratifiedKFold

x = dataFrame_justo.drop('Class', axis = 'columns')
y = dataFrame_justo['Class']


raw_train_x, raw_test_x, train_y, test_y = train_test_split(x, y, test_size = 0.25) #stratified não é necessário

model = LinearSVC()
cv = StratifiedKFold(n_splits=5, shuffle=True)
results = cross_validate(model, raw_train_x, train_y, cv = cv, return_train_score = False)
imprimeResult(results)

Resultados cross validation: [0.64864865 0.48648649 0.49324324 0.54421769 0.6122449 ]
Média da precisão: 55.70 %
Desvio padrão: 6.43 %
Acurracy esta entre: [ 42.84320168097086 --- 68.55043685552481 ]


<h1>SVC() com dados justos</h1>

In [88]:
#Escalando as dimensões de análise do modelo usando pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.pipeline import Pipeline
from sklearn.model_selection import cross_validate
from sklearn.model_selection import StratifiedKFold

SEED = 301
np.random.seed(SEED)

scaler = StandardScaler()
model = SVC()

pipeline = Pipeline([('scaler', scaler), ('model', model)])

cv = StratifiedKFold(n_splits = 5, shuffle = True)

#AO invés do modelo, aqui é passado o pipeline
result = cross_validate(pipeline, x, y, cv = cv)

imprimeResult(result)

Resultados cross validation: [0.95939086 0.93908629 0.91878173 0.92893401 0.91836735]
Média da precisão: 93.29 %
Desvio padrão: 1.53 %
Acurracy esta entre: [ 90.23665007431694 --- 96.3457595392747 ]


<h1>DECISION TREE com dados justos</h1>

In [89]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_validate
from sklearn.model_selection import StratifiedKFold

model = DecisionTreeClassifier()

cv = StratifiedKFold(n_splits = 5, shuffle = True)

result = cross_validate(model, x, y, cv = cv)

imprimeResult(result)

Resultados cross validation: [0.92385787 0.89340102 0.8680203  0.89847716 0.91326531]
Média da precisão: 89.94 %
Desvio padrão: 1.90 %
Acurracy esta entre: [ 86.13311515599345 --- 93.74775089601111 ]
